compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala

```diff
@@ -46,14 +61,115 @@ class ExtractSemanticDB extends Phase:
+  override def runOn(units: List[CompilationUnit])(using ctx: Context): List[CompilationUnit] = {
+    val appendDiagnostics = phaseMode == ExtractSemanticDB.PhaseMode.AppendDiagnostics
+    val warnings =
+      if (appendDiagnostics)
+        ctx.reporter.allWarnings.groupBy(w => w.pos.source)
+      else Map.empty
+
+    units.asJava.parallelStream().forEach { unit =>
+      val unitCtx = ctx.fresh.setCompilationUnit(unit).withRootImports
+      if (appendDiagnostics)
+        warnings.get(unit.source).foreach { ws =>
+          ExtractSemanticDB.appendDiagnostics(unit.source, ws.map(_.toSemanticDiagnostic))
+        }
+      else
+        val extractor = ExtractSemanticDB.Extractor()
+        extractor.extract(unit.tpdTree)
+        ExtractSemanticDB.write(unit.source, extractor.occurrences.toList, extractor.symbolInfos.toList, extractor.synthetics.toList)
+    }
+    units
+  }
+
+  def run(using Context): Unit = unsupported("run")
+end ExtractSemanticDB
+
+object ExtractSemanticDB:
+  import java.nio.file.Path
+  import java.nio.file.Files
+  import java.nio.file.Paths
+
+  val phaseNamePrefix: String = "extractSemanticDB"
+  val description: String = "extract info into .semanticdb files"
+
+  enum PhaseMode:
+    case ExtractSemanticInfo
+    case AppendDiagnostics
+
+  class ExtractSemanticInfo extends ExtractSemanticDB(PhaseMode.ExtractSemanticInfo, "ExtractSemanticInfo")
+
+  class AppendDiagnostics extends ExtractSemanticDB(PhaseMode.AppendDiagnostics, "AppendDiagnostics")
+
+  private def semanticdbTarget(using Context): Option[Path] =
+    Option(ctx.settings.semanticdbTarget.value)
+      .filterNot(_.isEmpty)
+      .map(Paths.get(_))
+
+  private def outputDirectory(using Context): AbstractFile = ctx.settings.outputDir.value
+
+  private def absolutePath(path: Path): Path = path.toAbsolutePath.normalize
+
+  private def write(
+    source: SourceFile,
+    occurrences: List[SymbolOccurrence],
+    symbolInfos: List[SymbolInformation],
+    synthetics: List[Synthetic],
+  )(using Context): Unit =
+    val outpath = semanticdbPath(source)
+    Files.createDirectories(outpath.getParent())
+    val doc: TextDocument = TextDocument(
+      schema = Schema.SEMANTICDB4,
+      language = Language.SCALA,
+      uri = Tools.mkURIstring(Paths.get(relPath(source))),
+      text = "",
+      md5 = internal.MD5.compute(String(source.content)),
+      symbols = symbolInfos,
+      occurrences = occurrences,
+      synthetics = synthetics,
+    )
+    val docs = TextDocuments(List(doc))
+    val out = Files.newOutputStream(outpath)
+    try
+      val stream = internal.SemanticdbOutputStream.newInstance(out)
+      docs.writeTo(stream)
+      stream.flush()
+    finally
+      out.close()
+  end write
+
+  private def appendDiagnostics(
+    source: SourceFile,
+    diagnostics: Seq[Diagnostic]
+  )(using Context): Unit =
```

Comment:
I also think this should not depend on Context (fix `semanticdbPath` first)
```suggestion
  ): Unit =
```

Reply:
We need this for `semanticdbPath`, should we use `settings` instead? (same as above)